dll/win32/shell32/CDefViewUtil.cpp

```diff
@@ -0,0 +1,132 @@
+        }
+};
+
+// This class adapts the legacy function callback to work as an IShellFolderViewCB
+class CShellFolderViewCBWrapper :
+    public CComObjectRootEx<CComMultiThreadModelNoCS>,
+    public IShellFolderViewCB,
+    public CObjectWithSiteBase
+{
+    protected:
+        HWND                    m_hWndMain;
+        CComPtr<IShellFolder>   m_psf;
+        CComPtr<IShellView>     m_psvOuter;
+        LPFNVIEWCALLBACK        m_Callback;
+        FOLDERVIEWMODE          m_FVM;
+        LONG                    m_Events;
+    public:
+        CShellFolderViewCBWrapper() : m_hWndMain(NULL) {}
+
+        HRESULT WINAPI Initialize(LPCSFV psvcbi)
+        {
+            m_psf = psvcbi->pshf;
+            m_psvOuter = psvcbi->psvOuter;
+            m_Callback = psvcbi->pfnCallback;
+            m_FVM = psvcbi->fvm;
+            m_Events = psvcbi->lEvents;
+            return S_OK;
+        }
+
+        // IShellFolderViewCB
+        STDMETHODIMP MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam) override
+        {
+            switch (uMsg)
+            {
+                case SFVM_HWNDMAIN:
+                    m_hWndMain = (HWND)lParam;
+                    break;
+
+                case SFVM_DEFVIEWMODE:
+                    if (m_FVM)
+                        *(FOLDERVIEWMODE*)lParam = m_FVM;
+                    break;
+            }
+
+            HRESULT hr = m_Callback(m_psvOuter, m_psf, m_hWndMain, uMsg, wParam, lParam);
+            if (SUCCEEDED(hr))
+                return hr;
+
+            switch (uMsg)
+            {
+                case SFVM_GETNOTIFY:
+                    *(LPITEMIDLIST*)wParam = NULL; // TODO: Where is this pidl supposed to come from?
+                    *(LONG*)lParam = m_Events;
+                    return S_OK;
+            }
+            return hr;
+        }
+
+        // IObjectWithSite
+        STDMETHODIMP SetSite(IUnknown *pUnkSite) override
+        {
+            //learn.microsoft.com/en-us/windows/win32/shell/sfvm-setisfv
+            HRESULT hr = CObjectWithSiteBase::SetSite(pUnkSite);
+            MessageSFVCB(SFVM_SETISFV, 0, (LPARAM)pUnkSite);
+            return hr;
+        }
+
+        DECLARE_NO_REGISTRY()
+        DECLARE_NOT_AGGREGATABLE(CShellFolderViewCBWrapper)
+        BEGIN_COM_MAP(CShellFolderViewCBWrapper)
+            COM_INTERFACE_ENTRY_IID(IID_IShellFolderViewCB, IShellFolderViewCB)
+            COM_INTERFACE_ENTRY_IID(IID_IObjectWithSite, IObjectWithSite)
+        END_COM_MAP()
+};
+
+/*************************************************************************
+ * SHCreateShellFolderViewEx [SHELL32.174] (Win95+)
+ */
+EXTERN_C HRESULT WINAPI
+SHCreateShellFolderViewEx(_In_ LPCSFV pcsfv, _Out_ IShellView **ppsv)
+{
+    if (!ppsv)
+        return E_INVALIDARG;
+    *ppsv = NULL;
+
+    TRACE("sf=%p pidl=%p cb=%p mode=0x%08x outer=%p\n",
+          pcsfv->pshf, pcsfv->pidl, pcsfv->pfnCallback,
+          pcsfv->fvm, pcsfv->psvOuter);
```

Comment:
Is the trace necessary? (I know I know, it's from the old code ^^)

Reply:
Is any function parameter trace necessary? I have another PR that causes the most common caller of this function to not call this anymore so the number of calls should be greatly reduced. Ideally only old 3rd-party code should be calling.

Reply:
as you wish ^^ i know code originally coming from wine has the habit of having such traces at the beginning, but there's a feature called "relay tracing" that could be enabled if this was _really_ needed.